File: Data Structures & Algorithms/valid-palindrome-ii/submission-0.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        return s == s[::-1]

    def validPalindrome(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        while l < r:
            if s[l] != s[r]:
                # delete s[l] or s[r]
                return self.isPalindrome(s[l + 1 : r + 1]) or self.isPalindrome(s[l:r])
            l += 1
            r -= 1
        return True

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0, head);
        int size = 0;
        ListNode cur = head;
        while (cur != null) {
            size++;
            cur = cur.next;
        }
        // traverse to predecessor (size - n) steps from dummy
        ListNode p = dummy;
        for (int i = 0; i < size - n; i++) {
            p = p.next;
        }
        // delete target
        p.next = p.next.next;
        return dummy.next;
    }
}
